HTTP streaming client, and trailer support

sunfishcode opened the pull request:
Add a wstd API for creating proxy applications, wrapping the WASI proxy world trait and macro.

Compiling the example with `RUSTFLAGS="-Clink-arg=--wasi-adapter=proxy"` produces a program that runs in `wasmtime serve`.

Marking as a draft for now, as it's still experimental. Feedback on the approach or "is this even a reasonable idea" welcome :smile:.

yoshuawuyts commented:
Thank you for submitting this! In general, I'm a big fan of making people's lives easier, and I believe that the general thrust of what you're filing here is the right thing to do. So, yes - I believe this is reasonable!

On a first pass, this is what stood out to me:

1. I quite enjoy the pattern Rust has for `main` being the main entry point of the application. While I know Components blur the boundary between libraries and binaries since everything is virtualizable, I feel like it would be especially nice here if `fn main` for HTTP proxies could Just Work? How would you feel about changing that?
2. There is a fair bit of mapping code inside of the macro happening. Could we turn those into conversion trait impls on the types instead? That keeps as much code out of the macros as possible.
3. Is `fn main(req) -> res` the right pattern we want to implement? I believe that might lead to some trouble handling HTTP trailers?

sunfishcode commented:
>     1. I quite enjoy the pattern Rust has for `main` being the main entry point of the application. While I know Components blur the boundary between libraries and binaries since everything is virtualizable, I feel like it would be especially nice here if `fn main` for HTTP proxies could Just Work? How would you feel about changing that?

Do you still like calling it `main` if it can be called multiple times on the same instance?

>     2. There is a fair bit of mapping code inside of the macro happening. Could we turn those into conversion trait impls on the types instead? That keeps as much code out of the macros as possible.

Yeah, I'll work on factoring that out.

>     3. Is `fn main(req) -> res` the right pattern we want to implement? I believe that might lead to some trouble handling HTTP trailers?

I think it'll lead to some trouble handling bodies too -}. I need to revamp the `Response`/`Request` types, and maybe even have separate types for the incoming-request side of things.

sunfishcode commented on src/http/proxy.rs:
```diff
@@ -0,0 +1,119 @@
+                let headers = wasi::http::types::Fields::new();
+                for (name, value) in response.headers() {
+                    let name = name.as_str().to_owned();
+                    let value = value.as_bytes().to_owned();
```
These `.to_owned()` calls are unnecessary with a wit-bindgen with https://github.com/bytecodealliance/wit-bindgen/pull/1102.

pchickey commented:
I was delayed a bit by starting a new job, but I've rewritten the reactor and streams abstractions in #35 #37. Hopefully that makes some aspects of this a bit simpler.

I'd like to avoid forcing users to ever use/know about RUSTFLAGS to compile applications. We can produce a HTTP Proxy world application with an ordinary `cargo build --target wasm32-wasip2` of a `bin` with a `wasi::http::proxy::export!(Wrapper)` long as there is a "dummy" `fn main() {}` for the linker to export as the cli world run, so lets make `#[wstd::proxy] async fn main(_: Request) -> Response {...}` work like that. I think in this framework we should err on the side of letting the common case be straightforward - the user wants to write a proxy, and doesn't care that it also happens to impl the CLI world's export in a vacuous way - rather than enabling complicated behavior, like the user that has both a CLI world run export and a proxy export. Sophisticated users can use other methods to get sophisticated behavior, such as using the `wasi` crate directly and passing RUSTFLAGS if they want.

sunfishcode commented:
I've now overhauled the API here.
 - It now uses `fn main` syntax for the entrypoint.
 - All the conversion code is now factored out of the macro.
 - I switched from the response return type to a `Resonder` argument, which allows it to support trailers as well as streaming bodies.
 - I added a placeholder `fn main` to make the crate compile with wasm32-wasip2 with no special `RUSTFLAGS`.

Here's a simple hello-world example:
```rust
#[wstd::proxy]
async fn main(request: Request<IncomingBody>, responder: Responder) -> Finished {
    responder
        .respond(Response::new(b"Hello, wasi:http/proxy world!\n"), None)
        .await
}
```
For more interesting examples, including a streaming body, see the examples/proxy.rs in the PR.

>  think in this framework we should err on the side of letting the common case be straightforward - the user wants to write a proxy, and doesn't care that it also happens to impl the CLI world's export in a vacuous way - rather than enabling sophisticated behavior, like the user that has both a CLI world run export and a proxy export. 

I've now implemented this. That said, a downside is that it's not just the export, it's also the imports, and it means that the resulting proxy components don't run in plain `wasmtime serve`; they need `wasmtime serve -S cli`.

pchickey commented on src/http/body.rs:
```diff
@@ -113,6 +113,12 @@ impl Body for Empty {
+        None
+    }
+}
+
```
Rather than implement this special case, could we export a struct BodyStream<S: AsyncRead> that gives an impl Body with len always None?

sunfishcode replied:
I've now made this change.

sunfishcode commented:
With other features factored out into separate PRs that have now landed, this PR now contains a series of patches adding output-streaming client and client trailers support, and a nice example.

pchickey commented on src/http/client.rs:
```diff
@@ -76,6 +137,25 @@ impl Client {
+/// to wait for the response.
+pub struct FutureIncomingResponse(WasiFutureIncomingResponse);
+
+impl FutureIncomingResponse {
```
I think this can be https://doc.rust-lang.org/std/future/trait.IntoFuture.html ?

sunfishcode replied:
I've now updated it to use `impl Future`.

pchickey commented on src/http/client.rs:
```diff
@@ -83,24 +84,35 @@ impl Client {
             fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-                match pin_project(self.subscription).poll(cx) {
+                let this = self.project();
+                match this.subscription.as_pin_mut().expect("make it so").poll(cx) {
```
Why is this option / expect here? Other than that, good to go!

sunfishcode replied:
That was just me being confused. The subsequent patch fixes that now.

pchickey commented on src/http/client.rs:
```diff
@@ -46,6 +63,90 @@ impl Client {
+        let outgoing_body = OutgoingBody::new(AsyncOutputStream::new(wasi_stream), wasi_body);
+
+        pin_project! {
+            struct IncomingResponseFuture {
```
```suggestion
            #[must_use = "futures do nothing unless polled or .awaited"]
            struct IncomingResponseFuture {
```
